### PR TITLE
refactor(onboarding): hoist login state into OnboardingViewModel ♻️

### PR DESCRIPTION
**Pull request by Klazomenai**
## Summary

- Hoists login async logic from `OnboardingActivity` into `OnboardingViewModel` with a `LoginState` sealed class (`Idle` / `InProgress` / `Error` / `Success`), decoupling coroutine lifecycle from the Activity and making state rotation-safe
- Opens `DeckChatApplication.secureStorage` for test injection; unit tests bypass the Tink/Keystore constructor via `PlaintextTokenEncryptor`
- Adds `testOptions.isIncludeAndroidResources = true` to Gradle so Robolectric binary resources load correctly

## Test plan

- [x] `OnboardingViewModelTest` — 5 unit tests: Idle initial state, InProgress→Success happy path, Error on failure, rotation semantics (StateFlow replay to new collector), double-submit guard while InProgress
- [x] `OnboardingActivityTest` — 11 Robolectric tests:
  - **#211 baseline (7)** — cold start lands on STEP_LOGIN; 500ms cold-start budget; rotation preserves currentStep on each of STEP_LOGIN / STEP_VOICE / STEP_PERMISSIONS / STEP_BLUETOOTH; first-step back-press finishes activity
  - **#203 in-flight + error rotation (2)** — rotation during InProgress preserves the spinner UI and `loginAction` is not re-fired across the recreate; rotation during Error preserves the error text on the recreated Activity (VM survives via `NonConfigurationInstances`)
  - **`OnboardingViewModel.Factory` (2)** — Factory uses `TestDeckChatApplication.secureStorage`; Factory rejects unrelated ViewModel classes
- [x] All 16 unit test suites pass locally (181 tests, 0 failures, 0 errors)

### Issue #203 (6 ACs)

| # | AC | Status |
|---|---|---|
| 1 | `OnboardingViewModel` exists with `loginState: StateFlow<LoginState>` | ✅ full |
| 2 | `STEP_LOGIN` collects from the flow via `setupLoginStateObserver()` | ✅ full |
| 3 | Rotation during `InProgress` preserves state — no leak, no re-fire (Activity-level test verifies UI + `loginCallCount == 1`) | ✅ full |
| 4 | `Error` state survives rotation (Activity-level test verifies UI text + retained state) | ✅ full |
| 5 | Robolectric test (#211 baseline) covers rotation across each onboarding step (4 per-step rotation tests) | ✅ full |
| `AC-#203.5` | Factory reads from `(application as DeckChatApplication).secureStorage`, not direct construction | ✅ full |

### Issue #211 (5 ACs)

| # | AC | Status |
|---|---|---|
| 1 | `OnboardingActivityTest` exists and runs | ✅ full |
| 2 | All 4 baseline cases pass (cold-start, advance through each step, rotation across each step, back-press reverses one step) | ⚠️ **partial** — see breakdown below |
| 3 | Class is structured for extension (per-concern tests, `preloadTestViewModel()` helper, `@Before`/`@After` plumbing) | ✅ full |
| 4 | Documented in #209's `docs/testing.md` | ❌ owned by #209 |
| `AC-#211.5` | Cold-start completes within 500ms budget (`elapsedMs <= 500`) | ⚠️ **partial** — caveat below |

#### AC-#211.2 baseline breakdown

| Scope-defined baseline case | Coverage |
|---|---|
| Cold-start lands on STEP_LOGIN | ✅ `cold start lands on STEP_LOGIN` |
| Advance through each step | ⚠️ **untested as click-through.** The rotation tests reach each step via `Bundle.putInt("current_step", N)` (savedInstanceState injection), so each step's view is verified visible, but the `goNext()` → `showStep()` advance logic itself is not exercised by clicking `btnNext`. Follow-up to be filed unless folded into #226. |
| Rotation across each step preserves currentStep | ✅ 4 per-step rotation tests |
| Back-press reverses one step | ⚠️ only the first-step finish path is covered (`currentStep == 0` → activity finishes). Multi-step reversal (`currentStep > 0` → decrement, Activity stays alive) is deferred to #226 (operator decision, [discussion-r3259811281](https://github.com/Klazomenai/deck-chat/pull/243#discussion_r3259811281)). |

#### AC-#211.5 caveat

The 500ms timer starts inside `Robolectric.buildActivity(...)`, after `RuntimeEnvironment.getApplication()` has already initialised the Application. So the measurement window covers **Activity creation + layout inflation only** — `Application.onCreate()` (which is what AC-#211.5's stated rationale targets, since `MigrationGate.migrateIfNeeded()` runs there) is **not** in the budget. The AC's literal text ("Activity.onCreate → STEP_LOGIN visible") matches the test, but the rationale's intent (defending B4 from MigrationGate runtime in Application.onCreate) is not defended by this test. A full-process-boot measurement is a future follow-up.

Refs #203
Refs #211

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: app/src/main/java/dev/klazomenai/deckchat/OnboardingActivity.kt [hunk 1]
```diff
@@ -58,10 +55,14 @@ class OnboardingActivity : AppCompatActivity() {
...
+        viewModel = ViewModelProvider(this, OnboardingViewModel.Factory(application))
+            .get(OnboardingViewModel::class.java)
+
         if (savedInstanceState != null) {
             currentStep = savedInstanceState.getInt(KEY_CURRENT_STEP, 0)
         }
 
+        setupLoginStateObserver()
         setupPermissionsStep()
         showStep(currentStep)
```
[the remark this answers is not in the document]
Comment (Klazomenai) [reply app/src/main/java/dev/klazomenai/deckchat/OnboardingActivity.kt]: Fixed in `6596b56` — `setupLoginStateObserver()` moved to after `showStep(currentStep)`. The race was: `showStep()` sets `btnNext.isEnabled = true` at line 92 unconditionally; if the observer ran first (replaying a retained `InProgress` state after rotation), `showStep` would immediately re-enable the button. Reversed order means `showStep` runs first, then the observer immediately overrides `isEnabled = false` for `InProgress`.

File: app/src/main/java/dev/klazomenai/deckchat/OnboardingViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+                _loginState.value = LoginState.Error(e.message ?: "Login failed (${e.javaClass.simpleName})")
+            }
+        }
+    }
+
+    class Factory(private val application: Application) : ViewModelProvider.Factory {
+        @Suppress("UNCHECKED_CAST")
+        override fun <T : ViewModel> create(modelClass: Class<T>): T {
+            val storage = (application as DeckChatApplication).secureStorage
+            return OnboardingViewModel(application, storage) as T
```
[the remark this answers is not in the document]
Comment (Klazomenai) [reply app/src/main/java/dev/klazomenai/deckchat/OnboardingViewModel.kt:71]: Fixed in `6596b56` — `require(modelClass.isAssignableFrom(OnboardingViewModel::class.java))` added to match `MainViewModel.Factory`'s pattern. A request for an unrelated ViewModel class now throws `IllegalArgumentException: Unknown ViewModel class: ...` rather than returning a silently mis-typed `OnboardingViewModel`.

File: app/src/test/java/dev/klazomenai/deckchat/OnboardingViewModelTest.kt [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        assertEquals(OnboardingViewModel.LoginState.Idle, vm.loginState.value)
+    }
+
+    @Test
+    fun `validateAndLogin transitions to InProgress then Success`() = runTest {
+        val vm = viewModel(loginAction = { _, _, _ -> /* no-op — instant success */ })
+
+        vm.validateAndLogin("https://matrix.example.com", "user", "pass", "!room:example.com")
+
+        assertEquals(OnboardingViewModel.LoginState.Success, vm.loginState.value)
```
[the remark this answers is not in the document]
Comment (Klazomenai) [reply app/src/test/java/dev/klazomenai/deckchat/OnboardingViewModelTest.kt]: Fixed in `6596b56`. The test now uses `StandardTestDispatcher(testScheduler)` as `ioDispatcher` and calls `advanceUntilIdle()`. With two `UnconfinedTestDispatcher` instances (main + io), `withContext(ioDispatcher)` is a no-op (same concrete dispatcher) and `StateFlow` conflation hides `InProgress` before the collector wakes up — the collector only ever sees `Success`. Using `StandardTestDispatcher` for io makes the context switch real: the ViewModel coroutine parks at `withContext`, the `UnconfinedTestDispatcher` collector processes `InProgress` eagerly, then `advanceUntilIdle()` runs the io task and `Success` follows.

File: app/src/test/java/dev/klazomenai/deckchat/OnboardingViewModelTest.kt [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    @Test
+    fun `loginState re-emits last value to new collector (rotation semantics)`() = runTest {
+        val vm = viewModel(loginAction = { _, _, _ -> throw RuntimeException("oops") })
+
+        vm.validateAndLogin("https://matrix.example.com", "user", "pass", "!room:example.com")
+
+        // New collector (simulates Activity recreation after rotation) sees the last state.
+        val collected = mutableListOf<OnboardingViewModel.LoginState>()
+        val state = vm.loginState.value
+        collected.add(state)
```
[the remark this answers is not in the document]
Comment (Klazomenai) [reply app/src/test/java/dev/klazomenai/deckchat/OnboardingViewModelTest.kt]: Fixed in `6596b56` — replaced the list-append pattern with `vm.loginState.first()`, a real suspension that subscribes to the Flow and waits for the first emitted value. `StateFlow` replays its retained value to every new collector, so a rotation regression (e.g. swapping `StateFlow` for a `MutableLiveData` or a one-shot `Channel`) would fail to emit and the test would time out rather than pass silently.

File: app/src/main/java/dev/klazomenai/deckchat/OnboardingViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        data class Error(val message: String) : LoginState()
+        object Success : LoginState()
+    }
+
+    private val _loginState = MutableStateFlow<LoginState>(LoginState.Idle)
+    val loginState: StateFlow<LoginState> = _loginState.asStateFlow()
+
+    fun validateAndLogin(url: String, username: String, password: String, roomId: String) {
+        viewModelScope.launch {
+            _loginState.value = LoginState.InProgress
```
[the remark this answers is not in the document]
Comment (Klazomenai) [reply app/src/main/java/dev/klazomenai/deckchat/OnboardingViewModel.kt:40]: Fixed in `6596b56` — `if (_loginState.value is LoginState.InProgress) return` added as the first line of `validateAndLogin()`. Concurrent calls while a login is in-flight are now silently dropped at the ViewModel level, regardless of whether the UI has disabled the button.

File: app/src/test/java/dev/klazomenai/deckchat/OnboardingActivityTest.kt [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+    /**
+     * Robolectric 4.16.1 + SDK 34 bug: PackageParser.generatePackageInfo returns an empty
+     * activities array, so the binary manifest theme is never stored in packageInfos.
+     * Register the activity with its declared theme before each test so ShadowActivity
+     * finds a non-zero themeResource and calls setTheme() during attach.
+     */
+    @Before
+    fun registerActivityTheme() {
+        val app = RuntimeEnvironment.getApplication()
```
[the remark this answers is not in the document]
Comment (Klazomenai) [reply app/src/test/java/dev/klazomenai/deckchat/OnboardingActivityTest.kt:103]: Acknowledged. The `buildActivity { }` call (where the timer starts) runs after `RuntimeEnvironment.getApplication()` has already initialised `DeckChatApplication`, so `Application.onCreate()` is not inside the measurement window. The budget therefore captures Activity creation + layout inflation only, not Application startup. #211 tracks adding a comment to the test that makes this boundary explicit, and optionally extending the measurement to cover the full cold-start path including Application init.

File: app/src/test/java/dev/klazomenai/deckchat/OnboardingActivityTest.kt [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+    }
+
+    @Test
+    fun `rotation on STEP_LOGIN preserves currentStep`() {
+        val controller = Robolectric.buildActivity(OnboardingActivity::class.java)
+            .create().start().resume()
+        controller.recreate()
+        val activity = controller.get()
+        val stepLogin = activity.findViewById<View>(R.id.step_login)
+        assertEquals(View.VISIBLE, stepLogin.visibility)
```
[the remark this answers is not in the document]
Comment (Klazomenai) [reply app/src/test/java/dev/klazomenai/deckchat/OnboardingActivityTest.kt:195]: Agreed — the rotation test here only covers the `Idle`-state path. A test that injects a suspended `loginAction`, saves/restores instance state, and verifies the new Activity renders `InProgress` or `Error` without re-triggering login is the real regression guard for this refactor. That test is deferred to #211 (rotation + error-state AC expansion).

File: app/src/test/java/dev/klazomenai/deckchat/OnboardingActivityTest.kt [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+    }
+
+    @Test
+    fun `back press on first step finishes activity`() {
+        val activity = Robolectric.buildActivity(OnboardingActivity::class.java)
+            .create().start().resume().get()
+        activity.onBackPressedDispatcher.onBackPressed()
+        assertTrue(activity.isFinishing)
```
[the remark this answers is not in the document]
Comment (Klazomenai) [reply app/src/test/java/dev/klazomenai/deckchat/OnboardingActivityTest.kt:242]: Agreed. The test only exercises the first-step finish path. Multi-step back-press (`currentStep > 0` should decrement and keep the Activity alive) is unexercised. Deferring to #226 where we plan Activity-level coverage expansion alongside the singleton-hoist work.

File: app/src/main/java/dev/klazomenai/deckchat/OnboardingViewModel.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (Klazomenai) [reply app/src/main/java/dev/klazomenai/deckchat/OnboardingViewModel.kt:71]: `DeckChatApplication.secureStorage` is constructed during `Application.onCreate()`, which runs once before any Activity is created. `Factory.create()` retrieves the already-initialised singleton via `(application as DeckChatApplication).secureStorage` — it does not construct a new `SecureStorage`. This PR therefore does not add any Keystore/Tink cost to `onboarding onCreate`. The lazy-provider / singleton-hoist question (moving the Factory lookup to an injected parameter rather than a direct `secureStorage` dereference) is tracked in #226.

File: app/src/test/java/dev/klazomenai/deckchat/OnboardingActivityTest.kt [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+ * 2. [registerActivityTheme]: Robolectric 4.16.1 + SDK 34 has a bug where
+ *    PackageParser.generatePackageInfo returns an empty activities array, so
+ *    packageInfos never receives ActivityInfo.theme from the binary manifest.
+ *    ShadowActivity.callAttach() then calls addActivityIfNotPresent() which inserts
+ *    a blank ActivityInfo (theme=0), meaning setTheme() is never called, and
+ *    AppCompatDelegateImpl.createSubDecor() throws "You need to use a Theme.AppCompat theme".
+ *    Fix: call shadowOf(pm).addOrUpdateActivity() in @Before to register the activity
+ *    with theme=R.style.Theme_DeckChat before any buildActivity() call.
+ *
+ * Covers #211 baseline ACs and the rotation/error state ACs from #203.
```
[the remark this answers is not in the document]
Comment (Klazomenai) [reply app/src/test/java/dev/klazomenai/deckchat/OnboardingActivityTest.kt]: Fixed in `dd8d8cf` — KDoc now lists the ACs actually covered (cold start, activity-creation budget, idle-state step restoration, first-step back-press) and explicitly names the deferred items: in-flight/error-state rotation → #211, multi-step back-press → #226.

File: app/src/main/java/dev/klazomenai/deckchat/OnboardingActivity.kt [hunk 10]
```diff
@@ -122,13 +123,43 @@ class OnboardingActivity : AppCompatActivity() {
...
+                        btnNext.isEnabled = false
+                    }
+                    is OnboardingViewModel.LoginState.Error -> {
+                        loginProgress.visibility = View.GONE
+                        loginError.text = state.message
+                        loginError.visibility = View.VISIBLE
+                        btnNext.isEnabled = true
+                    }
+                    is OnboardingViewModel.LoginState.Success -> {
+                        loginProgress.visibility = View.GONE
```
[the remark this answers is not in the document]
Comment (Klazomenai) [reply app/src/main/java/dev/klazomenai/deckchat/OnboardingActivity.kt]: Fixed in `dd8d8cf` — `loginError.visibility = View.GONE` added to the `Success` branch. Without it, a retry after an error (Error → InProgress → Success) would leave the stale error text visible even after the login step advances.

File: app/src/test/java/dev/klazomenai/deckchat/OnboardingViewModelTest.kt [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+
+        val states = mutableListOf<OnboardingViewModel.LoginState>()
+        val job = launch { vm.loginState.collect { states.add(it) } }
+
+        vm.validateAndLogin("https://matrix.example.com", "user", "pass", "!room:example.com")
+        advanceUntilIdle()
+        job.cancel()
+
+        assertTrue("InProgress must be emitted before Success", states.contains(OnboardingViewModel.LoginState.InProgress))
+        assertEquals(OnboardingViewModel.LoginState.Success, states.last())
```
[the remark this answers is not in the document]
Comment (Klazomenai) [reply app/src/test/java/dev/klazomenai/deckchat/OnboardingViewModelTest.kt]: Fixed in `dd8d8cf` — assertion now uses index comparison: `val inProgressIdx = states.indexOf(InProgress); val successIdx = states.lastIndexOf(Success); assertTrue(inProgressIdx in 0 until successIdx)`. A Success-before-InProgress-before-Success sequence would give `inProgressIdx > 0` but `successIdx = 0` (from `lastIndexOf`), failing the `inProgressIdx < successIdx` check. The `states.last()` assertion is kept as the terminal-state check.

File: app/src/main/java/dev/klazomenai/deckchat/OnboardingViewModel.kt [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+        object Success : LoginState()
+    }
+
+    private val _loginState = MutableStateFlow<LoginState>(LoginState.Idle)
+    val loginState: StateFlow<LoginState> = _loginState.asStateFlow()
+
+    fun validateAndLogin(url: String, username: String, password: String, roomId: String) {
+        if (_loginState.value is LoginState.InProgress) return
+        viewModelScope.launch {
+            _loginState.value = LoginState.InProgress
```
[the remark this answers is not in the document]
Comment (Klazomenai) [reply app/src/main/java/dev/klazomenai/deckchat/OnboardingViewModel.kt:40]: Fixed in `dd8d8cf` — added `second validateAndLogin call while InProgress is dropped` test. It uses a `CompletableDeferred` as the loginAction so the first login parks at `loginProceed.await()` after `advanceUntilIdle()`. The second call is made while `loginState == InProgress`; the test asserts state is still `InProgress` (no new coroutine launched), then completes the deferred and verifies only the first call's `roomId` was written.

File: app/src/main/java/dev/klazomenai/deckchat/OnboardingActivity.kt [hunk 13]
```diff
@@ -122,13 +123,44 @@ class OnboardingActivity : AppCompatActivity() {
...
+                        loginError.visibility = View.GONE
+                        if (currentStep == STEP_LOGIN) advanceStep()
```
[the remark this answers is not in the document]
Comment (Klazomenai) [reply app/src/main/java/dev/klazomenai/deckchat/OnboardingActivity.kt]: Fixed in `f73b1ac` — `acknowledgeSuccess()` added to `OnboardingViewModel`: resets `Success → Idle` after the Activity navigates past STEP_LOGIN. The Activity calls it immediately after `advanceStep()` inside the `Success` branch. On rotation or any subsequent Activity recreation, the StateFlow replays `Idle`, so the guard `if (currentStep == STEP_LOGIN)` is never re-triggered. The roomId is already persisted in `SecureStorage` at this point, so resetting to `Idle` loses no durable state.

File: app/src/main/java/dev/klazomenai/deckchat/OnboardingActivity.kt [hunk 14]
```diff
@@ -122,13 +123,44 @@ class OnboardingActivity : AppCompatActivity() {
 
     // --- Step 1: Login ---
 
+    private fun setupLoginStateObserver() {
+        val loginProgress = findViewById<ProgressBar>(R.id.login_progress)
+        val loginError = findViewById<TextView>(R.id.login_error)
+
+        lifecycleScope.launch {
+            viewModel.loginState.collect { state ->
```
[the remark this answers is not in the document]
Comment (Klazomenai) [reply app/src/main/java/dev/klazomenai/deckchat/OnboardingActivity.kt]: Fixed in `f73b1ac` — collector is now wrapped in `repeatOnLifecycle(Lifecycle.State.STARTED)`, matching `MainActivity`'s pattern. The collector is cancelled when the Activity stops and relaunched when it starts again; `StateFlow` replay restores the latest state to the new collector on re-entry, so no state is missed while respecting the visible lifecycle.

File: app/src/main/java/dev/klazomenai/deckchat/OnboardingViewModel.kt [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    class Factory(private val application: Application) : ViewModelProvider.Factory {
+        @Suppress("UNCHECKED_CAST")
+        override fun <T : ViewModel> create(modelClass: Class<T>): T {
+            require(modelClass.isAssignableFrom(OnboardingViewModel::class.java)) {
+                "Unknown ViewModel class: ${modelClass.name}"
+            }
+            val storage = (application as DeckChatApplication).secureStorage
+            return OnboardingViewModel(application, storage) as T
```
[the remark this answers is not in the document]
Comment (Klazomenai) [reply app/src/main/java/dev/klazomenai/deckchat/OnboardingViewModel.kt:71]: Fixed in `b540d2d` — two Factory tests added to `OnboardingActivityTest` (where `TestDeckChatApplication` with `PlaintextTokenEncryptor` is already configured): `Factory creates ViewModel using TestDeckChatApplication secureStorage` verifies `Factory(application).create(OnboardingViewModel::class.java)` succeeds and returns a ViewModel with `loginState == Idle` (implicitly verifying the `(application as DeckChatApplication).secureStorage` path works with the injected storage); `Factory throws for unrelated ViewModel class` verifies the `isAssignableFrom` guard throws `IllegalArgumentException`.

File: app/src/test/java/dev/klazomenai/deckchat/OnboardingViewModelTest.kt [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+
+        vm.validateAndLogin("https://matrix.example.com", "user", "pass", "!room:example.com")
+        advanceUntilIdle()
+        job.cancel()
+
+        val inProgressIdx = states.indexOf(OnboardingViewModel.LoginState.InProgress)
+        val successIdx = states.lastIndexOf(OnboardingViewModel.LoginState.Success)
+        assertTrue("InProgress must appear before Success in state sequence",
+            inProgressIdx in 0 until successIdx)
+        assertEquals(OnboardingViewModel.LoginState.Success, states.last())
```
[the remark this answers is not in the document]
Comment (Klazomenai) [reply app/src/test/java/dev/klazomenai/deckchat/OnboardingViewModelTest.kt:86]: Fixed in `b540d2d` — changed from `lastIndexOf(Success)` to `indexOf(Success)` (first occurrence). A broken sequence like `[Idle, Success, InProgress, Success]` now gives `firstSuccessIdx = 1` and `inProgressIdx = 2`, so `2 in 0 until 1` is false and the assertion fails. Also added an inline comment explaining why `indexOf` (not `lastIndexOf`) is required.

File: app/src/main/java/dev/klazomenai/deckchat/DeckChatApplication.kt [hunk 17]
```diff
@@ -15,15 +15,15 @@ import kotlinx.coroutines.runBlocking
...
+open class DeckChatApplication : Application() {
 
     /**
      * App-wide singleton SecureStorage. Lazy so the Tink keyset load + Keystore round-trip
      * happens once and is shared across all callers. Call-site refactor (rewriting the five
      * dispersed construction sites to use this property) lands in the sibling PR for #226 —
      * this PR adds the property only.
      */
-    val secureStorage: SecureStorage by lazy { SecureStorage(applicationContext) }
+    open val secureStorage: SecureStorage by lazy { SecureStorage(applicationContext) }
```
[the remark this answers is not in the document]
Comment (Klazomenai) [reply app/src/main/java/dev/klazomenai/deckchat/DeckChatApplication.kt:29]: Fixed in `b540d2d` — KDoc updated to reflect current state: describes the singleton as shared across MainActivity, SettingsActivity, OnboardingActivity, and their ViewModels, and documents the `open` declaration as the test-injection seam (override with `PlaintextTokenEncryptor` to avoid the Android Keystore dependency in JVM tests). All stale rollout-sequencing text removed.

File: app/src/test/java/dev/klazomenai/deckchat/OnboardingActivityTest.kt [hunk 18]
```diff
@@ -0,0 +1,321 @@
...
+
+    @Test
+    fun `cold start STEP_LOGIN visible within 500ms`() {
+        val start = System.nanoTime()
+        val activity = Robolectric.buildActivity(OnboardingActivity::class.java)
+            .create().start().resume().get()
+        val stepLogin = activity.findViewById<View>(R.id.step_login)
+        val elapsedMs = (System.nanoTime() - start) / 1_000_000
+        assertEquals(View.VISIBLE, stepLogin.visibility)
+        assertTrue("onCreate exceeded 500ms budget (${elapsedMs}ms)", elapsedMs < 500)
```
[the remark this answers is not in the document]
Comment (Klazomenai) [reply app/src/test/java/dev/klazomenai/deckchat/OnboardingActivityTest.kt]: Fixed in `0fbafbd` — `elapsedMs < 500` is now `elapsedMs <= 500` so the boundary matches the "within 500ms" budget literally (a measurement of exactly 500ms passes, as stated).

File: app/src/test/java/dev/klazomenai/deckchat/OnboardingActivityTest.kt [hunk 19]
```diff
@@ -0,0 +1,321 @@
...
+
+        val app = RuntimeEnvironment.getApplication()
+        val activityInfo = ActivityInfo().apply {
+            packageName = app.packageName
+            name = OnboardingActivity::class.java.name
+            theme = R.style.Theme_DeckChat
+        }
+        val shadowPm = shadowOf(app.packageManager)
+        shadowPm.addOrUpdateActivity(activityInfo)
+
```
[the remark this answers is not in the document]
Comment (Klazomenai) [reply app/src/test/java/dev/klazomenai/deckchat/OnboardingActivityTest.kt:118]: Fixed in `0fbafbd` — `setUp()` now calls `app.getSharedPreferences("test_onboarding", Context.MODE_PRIVATE).edit().clear().commit()` before any `buildActivity()` runs. The point of the leak is real: `OnboardingViewModel.validateAndLogin` writes `storage.roomId = roomId` on Success, and that write is durable in the test prefs across methods unless we wipe them in `@Before`. The current PR's Activity tests don't reach Success (the rotation tests park at InProgress or throw to Error, the Factory tests never call `validateAndLogin`), but the clear is added defensively so that any future Success-path test won't silently contaminate the suite.

File: app/src/test/java/dev/klazomenai/deckchat/OnboardingActivityTest.kt [hunk 20]
```diff
@@ -0,0 +1,321 @@
...
+     * Must be called between `Robolectric.buildActivity()` and `controller.create()`.
+     */
+    private fun preloadTestViewModel(
+        activity: OnboardingActivity,
+        loginAction: suspend (String, String, String) -> Unit,
+    ): OnboardingViewModel {
+        val testVm = OnboardingViewModel(
+            activity.application,
+            (activity.application as DeckChatApplication).secureStorage,
+            ioDispatcher = Dispatchers.Unconfined,
```
[the remark this answers is not in the document]
Comment (Klazomenai) [reply app/src/test/java/dev/klazomenai/deckchat/OnboardingActivityTest.kt]: Fixed in `0fbafbd` — `preloadTestViewModel()` now passes the class-level `testDispatcher` (`UnconfinedTestDispatcher()`) as the ViewModel's `ioDispatcher` instead of the global `Dispatchers.Unconfined`. The same dispatcher backs `Dispatchers.Main` via `setMain(testDispatcher)`, so the `viewModelScope.launch` and the inner `withContext(io)` are both driven by test-controlled scheduling — the production code's dispatcher-switching contract is now exercised under a single test scheduler instead of partially escaping into the global `Unconfined` dispatcher.

File: app/src/test/java/dev/klazomenai/deckchat/OnboardingActivityTest.kt [hunk 21]
```diff
@@ -0,0 +1,331 @@
...
+ *    in Robolectric's binary-resource loader.
+ *
+ * 2. [registerActivityTheme]: Robolectric 4.16.1 + SDK 34 has a bug where
+ *    PackageParser.generatePackageInfo returns an empty activities array, so
+ *    packageInfos never receives ActivityInfo.theme from the binary manifest.
+ *    ShadowActivity.callAttach() then calls addActivityIfNotPresent() which inserts
+ *    a blank ActivityInfo (theme=0), meaning setTheme() is never called, and
+ *    AppCompatDelegateImpl.createSubDecor() throws "You need to use a Theme.AppCompat theme".
+ *    Fix: call shadowOf(pm).addOrUpdateActivity() in @Before to register the activity
+ *    with theme=R.style.Theme_DeckChat before any buildActivity() call.
```
[the remark this answers is not in the document]
Comment (Klazomenai) [reply app/src/test/java/dev/klazomenai/deckchat/OnboardingActivityTest.kt]: Fixed in `ca2b37d` — `[registerActivityTheme]` updated to `[setUp]` (`@Before`). The function was renamed from `registerActivityTheme` → `setUp` when `Dispatchers.setMain(testDispatcher)` was folded into it (this PR's round-1 commit), but the KDoc cross-reference was not updated in the same edit. The link now resolves to the actual symbol.
